Chore/vite migration

mariojsnunes opened the pull request:
PR Checklist

- [x] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Developer experience (improves developer workflows for contributing to the project)

## Description

This PR replaces `webpack` and `CRA` in favor of `vite`.
Tried to keep `Jest` but had too many issues so migrated to `vitest`.
Had issues on all pipeline steps due to:
- commonjs/esm
- typescript errors
- functions still use webpack and commonjs, and were importing the new vite/esm packages
- observable decorators stopped working for some reason, had to migrate to the recommended approach
- loading interfaces/types from `oa-components` caused issues, so replaced with `any` as a temporary fix
- cypress configuration with vite
 
## Git Issues

Closes #

benfurber commented on packages/cypress/src/support/db/endpoints.ts:
```diff
@@ -2,7 +2,7 @@ import { generateDBEndpoints } from 'oa-shared'
 // React apps populate a process variable, however it might not always be accessible outside
 // (e.g. cypress will instead use it's own env to populate a prefix)
-const process = globalThis.process || ({} as any)
+// const process = globalThis.process || ({} as any)
```
What's the context to this change?

mariojsnunes replied:
apparently isn't needed, I put it back, doesn't seem to have an effect

benfurber reviewed (commented):
Gone through in order up until the common folder. Had some minor comments so thought I'd post them as I need to do other things for a bit.

mariojsnunes commented on functions/webpack.config.ts:
yes, functions still use webpack. can be done on a future PR

mariojsnunes commented on packages/components/src/VideoPlayer/VideoPlayer.tsx:
```diff
@@ -1,4 +1,4 @@
-import ReactPlayer from 'react-player/lazy'
+import ReactPlayer from 'react-player'
```
No, but lazy fails with vite. should be fixed with the 3.0 release
https://github.com/cookpete/react-player/issues/1443

mariojsnunes commented on packages/components/src/test/setup.ts:
```diff
@@ -1,9 +1,9 @@
-import matchers from '@testing-library/jest-dom/matchers'
+// import matchers from '@testing-library/jest-dom/matchers'
 import { cleanup } from '@testing-library/react'
-import { afterEach, expect } from 'vitest'
+import { afterEach } from 'vitest'
 
 // extends Vitest's expect method with methods from react-testing-library
-expect.extend(matchers)
+// expect.extend(matchers)
```
couldn't get it to work... got this error:
![image](https://github.com/ONEARMY/community-platform/assets/8073622/d34a488f-138c-403a-86ca-161ef9b19f5b)

I believe this is the reason `import '@testing-library/jest-dom/vitest'` had to be included on most test files.

mariojsnunes commented on packages/components/src/index.ts:
```diff
@@ -57,4 +57,5 @@ export { UserEngagementWrapper } from './UserEngagementWrapper/UserEngagementWra
 export { DonationRequest } from './DonationRequest/DonationRequest'
 export { DonationRequestModal } from './DonationRequestModal/DonationRequestModal'
-export { UserNotificationItem } from './NotificationItem/NotificationItem'
+// export { IImageGalleryItem } from './ImageGallery/ImageGallery'
+// export { availableGlyphs } from './Icon/types'
```
got weird issues importing interfaces and types from the components library...
as a workaround found a way to not use them.

mariojsnunes commented on packages/components/src/TextNotification/TextNotification.tsx:
```diff
@@ -6,14 +6,14 @@ const fadeIn = keyframes({
 })
 
-export interface Props {
+export interface TextNotificationProps {
```
unlike others, this is exported so a better name is important.
(I'd prefer the convention to be ComponentNameProps)

davehakkens commented:
Can a PR like this break/change things on the frontend? 

-Asking because it seem to touch many areas and often after PR's like this many small functionalities stop working, the site is large, we dont test them all. Resulting into many bug reports and a bad user experience..

So this is me trying to be ahead of that🙃

mariojsnunes commented:
Forgot to add:
- changed library that renders country flags due to esm incompatibility and it wasn't maintained for 5 years
- also due to esm, updated the Uppy version and configured it's compressor (automatically optimizes images)

mariojsnunes commented on index.html:
```diff
@@ -0,0 +1,64 @@
+    <meta property="og:title" content="Community Platform" />
+    <meta
+      property="og:description"
+      content="A series of tools for the Precious Plastic community to collaborate around the world. Connect, share and meet each other to tackle plastic waste."
```
As I understand it, they need to exist and are changed before the deploy. See `scripts/postBuild.ts`

davehakkens commented:
When browsing 3 minutes (not going much into functionality test)
1. Blobs dont show correct
<img width="131" alt="Screenshot 2024-06-04 at 10 45 12" src="https://github.com/ONEARMY/community-platform/assets/13672737/0f2e7ba5-56ad-4b55-9521-d2cbfc4c86da">
<br>

2. profiles tabs dont show
<img width="408" alt="Screenshot 2024-06-04 at 10 45 51" src="https://github.com/ONEARMY/community-platform/assets/13672737/fa55de7d-58ae-4852-ad59-b9635b0e1ac8">
<br>

3. Notification got black instead of grey
<img width="120" alt="Screenshot 2024-06-04 at 10 46 13" src="https://github.com/ONEARMY/community-platform/assets/13672737/5c47f4de-a53e-493d-899f-1029271c1195">
<br>

4. Patreon connection doesnt work. (could be because its devsite?)
<img width="409" alt="Screenshot 2024-06-04 at 10 46 37" src="https://github.com/ONEARMY/community-platform/assets/13672737/f3c80d77-f686-4803-9569-9a35b8524193">
<br>

5. useful on how-to doesnt seem to work
<img width="234" alt="Screenshot 2024-06-04 at 10 48 24" src="https://github.com/ONEARMY/community-platform/assets/13672737/2dc4be91-81a0-4de9-b254-c39d090ab8aa">


Conclusion. I think we need more real testing before merging PR's like this.

mariojsnunes commented:
> **Conclusion. I think we need more real testing before merging PR's like this.**

Thanks for the tests!
Before merging I'll go over all the component changes again.

Have you tried image upload feature? Notice any quality difference?

As for the reported issues:
- I can replicate all locally except the 1.
- For 4. @benfurber can you confirm this would be due being the preview site?

mariojsnunes commented:
All those reported issues are fixed. (still need to confirm the Patreon issue)

Developer notes:
- using inline svgs now requires a double quote:
before: url(path-to-svg.svg)
after: url("path-to-svg.svg")

- using decorators no longer works (cause for .5):
![image](https://github.com/ONEARMY/community-platform/assets/8073622/0790ead0-8d99-4f06-9aca-a2f803b29377)
the issue was reported only for howtos, but the fix was applied to all stores.

davehakkens commented:
> Have you tried image upload feature? Notice any quality difference? @mariojsnunes  

Works nice for me.


Dont see any visual errors after a quick browse.
But should all functions also be tested? (thinking that the useful button didnt work before)

mariojsnunes commented:
> (thinking that the useful button didnt work before)

The cause for that was decorators used by our stores like @observable that stopped working with vite. They have now all been migrated to the new approach, so the risk is low.
